base/collection_xattr_test.go

```diff
@@ -1040,6 +1133,195 @@ func TestWriteUpdateWithXattrs(t *testing.T) {
+			xattrsToDelete: []string{"xattr1"},
+			cas:            uint64(1),
+			errorIs:        sgbucket.ErrUpsertAndDeleteSameXattr,
+		},
+		{
+			name:   "body=true,xattrs=xattr1,xattrsToDelete=nil,cas=0",
+			body:   []byte(`{"foo": "bar"}`),
+			xattrs: map[string][]byte{"xattr1": []byte(`{"a" : "b"}`)},
+		},
+		{
+			name:      "body=true,xattrs=xattr1,xattrsToDelete=nil,cas=1",
+			body:      []byte(`{"foo": "bar"}`),
+			xattrs:    map[string][]byte{"xattr1": []byte(`{"a" : "b"}`)},
+			cas:       uint64(1),
+			errorFunc: requireDocNotFoundOrCasMismatchError,
+		},
+		{
+			name:    "xattr_nil_value",
+			body:    []byte(`{"foo": "bar"}`),
+			xattrs:  map[string][]byte{"xattr1": nil},
+			errorIs: sgbucket.ErrNilXattrValue,
+		},
+	}
+	if bucket.IsSupported(sgbucket.BucketStoreFeatureMultiXattrSubdocOperations) {
+		tests = append(tests, []testCase{
+			{
+				name:           "body=true,xattrs=nil,xattrsToDelete=xattr1,xattr2,cas=0",
+				body:           []byte(`{"foo": "bar"}`),
+				xattrsToDelete: []string{"xattr1", "xattr2"},
+				errorIs:        sgbucket.ErrDeleteXattrOnDocumentInsert,
+			},
+			{
+				name:           "body=true,xattrs=nil,xattrsToDelete=xattr1,xattr2,cas=1",
+				body:           []byte(`{"foo": "bar"}`),
+				xattrsToDelete: []string{"xattr1", "xattr2"},
+				cas:            uint64(1),
+				errorFunc:      requireDocNotFoundOrCasMismatchError,
+			},
+			{
+				name:           "body=true,xattrs=xattr1,xattrsToDelete=xattr1,xattr2,cas=0",
+				body:           []byte(`{"foo": "bar"}`),
+				xattrs:         map[string][]byte{"xattr1": []byte(`{"a" : "b"}`)},
+				xattrsToDelete: []string{"xattr1", "xattr2"},
+				errorIs:        sgbucket.ErrDeleteXattrOnDocumentInsert,
+			},
+
+			{
+				name:           "body=true,xattrs=xattr1,xattrsToDelete=xattr1,xattr2,cas=1",
+				body:           []byte(`{"foo": "bar"}`),
+				xattrs:         map[string][]byte{"xattr1": []byte(`{"a" : "b"}`)},
+				xattrsToDelete: []string{"xattr1", "xattr2"},
+				cas:            uint64(1),
+				errorIs:        sgbucket.ErrUpsertAndDeleteSameXattr,
+			},
+		}...)
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.errorFunc != nil && test.errorIs != nil {
+				require.FailNow(t, "test case should specify errFunc  xor errorIs")
+			}
+			docID := t.Name()
+			cas, err := col.WriteWithXattrs(ctx, docID, 0, test.cas, test.body, test.xattrs, test.xattrsToDelete, nil)
+			if test.errorFunc != nil {
+				test.errorFunc(t, err)
+				require.Equal(t, uint64(0), cas)
+				return
+			}
+			require.ErrorIs(t, err, test.errorIs)
+			if test.errorIs != nil {
+				require.Equal(t, uint64(0), cas)
+				return
+			}
+			require.NoError(t, err)
+			require.NotEqual(t, uint64(0), cas)
+
+			xattrKeys := make([]string, 0, len(test.xattrs))
+			for k := range test.xattrs {
+				xattrKeys = append(xattrKeys, k)
+			}
+			xattrKeys = append(xattrKeys, test.xattrsToDelete...)
```

Comment:
I'm not sure why we're expecting to find xattrsToDelete here.  Running the test locally we don't have any subtests that reach this point where xattrsToDelete is non-nil.  Are we missing defining a successful case where WriteWithXattrs successfully deletes xattr(s)?

Reply:
It was a subtlety of earlier testing and copy/paste from a different test. We do successfully delete xattrs, but they all have to be written first, and so they are defined in `test.xattrs`